### PR TITLE
[SPARK-49980][CORE][SQL] Fix potential file stream leaks caused by interruption in canceled tasks

### PR DESCRIPTION
**Pull request by davintjong-db**
### What changes were proposed in this pull request?

Fix potential file stream leaks caused by interruption in canceled tasks.

### Why are the changes needed?

There is a common pattern of:

```
val intermediateReader = ... constructor ...
intermediateReader.initialize(...) <- this may throw an error, i.e. if the task is interrupted.
return WrapperIterator(intermediateReader)
```

If an error is thrown in `initialize`, `intermediateReader` is never closed and the caller has no reference and cannot close it itself, causing the open stream to leak.

This is handled in many other analogous places, but not everywhere. Fixing some of them here.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala [hunk 1]
```diff
@@ -244,7 +244,13 @@ class NewHadoopRDD[K, V](
         try {
           val _reader = format.createRecordReader(
             split.serializableHadoopSplit.value, hadoopAttemptContext)
-          _reader.initialize(split.serializableHadoopSplit.value, hadoopAttemptContext)
+          try {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala:247]: can we use `Utils.tryWithResource` instead?

Comment (davintjong-db) [reply core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala:247]: I don't think that has the right semantics: we only want to close on error, whereas I think `tryWithResource` will close at the end of the try block no matter what.

Maybe I can add another util to clean this up?

File: core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala [hunk 2]
```diff
@@ -46,9 +46,9 @@ import org.apache.spark.util.{SerializableConfiguration, ShutdownHookManager, Ut
 import org.apache.spark.util.ArrayImplicits._
 
 private[spark] class NewHadoopPartition(
-    rddId: Int,
-    val index: Int,
-    rawSplit: InputSplit with Writable)
+                                         rddId: Int,
+                                         val index: Int,
+                                         rawSplit: InputSplit with Writable)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala:51]: Let's remove those unrelated changes

File: common/utils/src/main/scala/org/apache/spark/util/SparkErrorUtils.scala [hunk 3]
```diff
@@ -48,6 +48,22 @@ private[spark] trait SparkErrorUtils extends Logging {
     try f.apply(resource) finally resource.close()
   }
 
+  /**
+   * Try to initialize a resource. If an exception is throw during initialization, closes the
+   * resource before propagating the error. Otherwise, the caller is responsible for closing
+   * the resource. This means that [[T]] should provide some way to close the resource.
+   */
+  def tryInitializeResource[R <: Closeable, T](createResource: => R)(initialize: R => T): T = {
+    val resource = createResource
```
Comment (jiangxb1987) [inline common/utils/src/main/scala/org/apache/spark/util/SparkErrorUtils.scala:57]: Why we don't need to capture the exception during creation, isn't it possible?

Comment (davintjong-db) [reply common/utils/src/main/scala/org/apache/spark/util/SparkErrorUtils.scala:57]: If an exception is thrown here:
```
val intermediateReader = ... constructor ... <- here
intermediateReader.initialize(...)
return WrapperIterator(intermediateReader)
```
It should be the responsibility of the constructor to close any resources it has opened if it fails, and the current code is not responsible for any open resources itself. So all we need to do in this case is propagate this error to the caller, which happens anyway.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: common/utils/src/main/scala/org/apache/spark/util/SparkErrorUtils.scala [hunk 4]
```diff
@@ -48,6 +48,22 @@ private[spark] trait SparkErrorUtils extends Logging {
...
+    try {
+      initialize(resource)
+    } catch {
+      case e: Throwable =>
+        resource.close()
+        throw e
```
Comment (mridulm) [inline common/utils/src/main/scala/org/apache/spark/util/SparkErrorUtils.scala:63]: If `close` results in an `IOException`, `e` will be lost - we should catch and add it via `addSuppressed`.
Do you want to create a follow up PR for this @davintjong-db ?
